addons/binding/org.openhab.binding.co2signal/src/main/java/org/openhab/binding/co2signal/handler/CO2SignalHandler.java

```diff
@@ -0,0 +1,292 @@
+    }
+
+    /**
+     * Request new CO2 Signal data to the co2signal.com service
+     *
+     * @param location geo-coordinates from config
+     * @param countryCode station ID from config
+     *
+     * @return the CO2 Signal data object mapping the JSON response or null in case of error
+     */
+    private CO2SignalJsonResponse getCO2SignalData(String location, String countryCode) {
+        CO2SignalJsonResponse result = null;
+        boolean resultOk = false;
+        String errorMsg = null;
+
+        try {
+            // Build a valid URL for the co2signal.com service
+            CO2SignalConfiguration config = getConfigAs(CO2SignalConfiguration.class);
+
+            String urlStr = URL;
+
+            if (location != null && !"".equals(location)) {
+                location = location.replace(" ", "").replace("\"", "").replace("'", "").trim();
+                String[] latLon = location.split(",");
+                urlStr = urlStr.replace("%QUERY%", "lat=" + latLon[0] + "&lon=" + latLon[1]);
+            } else {
+                urlStr = urlStr.replace("%QUERY%", "countryCode=" + countryCode);
+            }
+
+            logger.debug("URL = {}", urlStr);
+
+            // Run the HTTP request and get the JSON response from co2cn.org
+            URL url = new URL(urlStr);
+            HttpsURLConnection connection = (HttpsURLConnection) url.openConnection();
+            connection.setRequestProperty("User-Agent", "Mozilla/5.0");
+            connection.setRequestProperty("auth-token", config.apikey);
+
+            try {
+                String response = IOUtils.toString(connection.getInputStream());
+                logger.debug("co2Response = {}", response);
+
+                // Map the JSON response to an object
+                result = gson.fromJson(response, CO2SignalJsonResponse.class);
+            } finally {
+                IOUtils.closeQuietly(connection.getInputStream());
+            }
+
+            if (result == null) {
+                errorMsg = "no data returned";
+            } else if (result.getData() != null && result.getStatus() != "error") {
+                resultOk = true;
+            } else {
+                errorMsg = "missing data sub-object";
+            }
+
+            if (!resultOk) {
+                logger.warn("Error in co2cn.org (CO2 Signal) response: {}", errorMsg);
+            }
+        } catch (MalformedURLException e) {
+            errorMsg = e.getMessage();
+            logger.warn("Constructed url is not valid: {}", errorMsg);
+        } catch (JsonSyntaxException e) {
+            errorMsg = "Configuration is incorrect";
+            logger.warn("Error running co2cn.org (CO2 Signal) request: {}", errorMsg);
+        } catch (IOException | IllegalStateException e) {
+            errorMsg = e.getMessage();
+        }
+
+        // Update the thing status
+        if (resultOk) {
+            updateStatus(ThingStatus.ONLINE, ThingStatusDetail.NONE, "OK");
```

Comment:
Your status is already ONLINE; no need to set it again?

Reply:
Don't know much about things, things are the thing that confuses me the most about OH2 :-)

 So i don't know if i can just remove the code here or do need to do extra stuff. What about the following code on setting the thing to offline in an error case. Will the thing be automatically online again after the error goes away. What happens if the state of resultOk in 3 runs would be:

1. true
2. false
3. true